kernel/src/net/socket/unix/stream/listener.rs

```diff
@@ -99,81 +149,83 @@ impl BacklogTable {
         self.backlog_sockets.write().remove(addr_key);
     }
 }
 
-struct Backlog {
+pub(super) struct Backlog {
     addr: UnixSocketAddrBound,
     pollee: Pollee,
     backlog: AtomicUsize,
-    incoming_conns: Mutex<VecDeque<Connected>>,
+    incoming_conns: Mutex<Option<VecDeque<Connected>>>,
+    pauser: Arc<Pauser>,
 }
 
 impl Backlog {
-    fn new(addr: UnixSocketAddrBound, backlog: usize) -> Self {
+    fn new(addr: UnixSocketAddrBound, pollee: Pollee, backlog: usize, is_shutdown: bool) -> Self {
+        let incoming_sockets = if is_shutdown {
+            None
+        } else {
+            Some(VecDeque::with_capacity(backlog))
+        };
+
         Self {
             addr,
-            pollee: Pollee::new(IoEvents::empty()),
+            pollee,
             backlog: AtomicUsize::new(backlog),
-            incoming_conns: Mutex::new(VecDeque::with_capacity(backlog)),
+            incoming_conns: Mutex::new(incoming_sockets),
+            pauser: Pauser::new(),
         }
     }
 
     fn addr(&self) -> &UnixSocketAddrBound {
         &self.addr
     }
 
-    fn push_incoming(&self, client_addr: Option<UnixSocketAddrBound>) -> Result<Connected> {
-        let mut incoming_conns = self.incoming_conns.lock();
-
-        if incoming_conns.len() >= self.backlog.load(Ordering::Relaxed) {
-            return_errno_with_message!(
-                Errno::EAGAIN,
-                "the pending connection queue on the listening socket is full"
-            );
-        }
-
-        let (server_conn, client_conn) = Connected::new_pair(Some(self.addr.clone()), client_addr);
-        incoming_conns.push_back(server_conn);
-
-        self.pollee.add_events(IoEvents::IN);
+    fn pop_incoming(&self) -> Result<Connected> {
+        let mut locked_incoming_conns = self.incoming_conns.lock();
 
-        Ok(client_conn)
-    }
+        let Some(incoming_conns) = &mut *locked_incoming_conns else {
+            return_errno_with_message!(Errno::EINVAL, "the socket is shut down for reading");
+        };
 
-    fn pop_incoming(&self) -> Result<Connected> {
-        let mut incoming_conns = self.incoming_conns.lock();
         let conn = incoming_conns.pop_front();
         if incoming_conns.is_empty() {
             self.pollee.del_events(IoEvents::IN);
         }
+
+        drop(locked_incoming_conns);
+
+        if conn.is_some() {
+            self.pauser.resume_one();
+        }
+
         conn.ok_or_else(|| Error::with_message(Errno::EAGAIN, "no pending connection is available"))
     }
 
     fn set_backlog(&self, backlog: usize) {
-        self.backlog.store(backlog, Ordering::Relaxed);
+        let old_backlog = self.backlog.swap(backlog, Ordering::Relaxed);
+
+        if old_backlog < backlog {
+            self.pauser.resume_all();
+        }
+    }
+
+    fn shutdown(&self) {
+        let mut incoming_conns = self.incoming_conns.lock();
+
+        let value = core::mem::take(&mut *incoming_conns);
```

Comment:
Using `*incoming_conns = None;` should be ok here?

Reply:
I think it is okay. Just a little more time in the critical part (holding the `incoming_conns` lock).